### PR TITLE
Add first files of data pipeline

### PR DESCRIPTION
**Pull request by warenlg**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Shell neural completion - Work in Progress
+
+## Pase 1 : Build data pipeline
```
Comment (vmarkovtsev) [inline README.md]: Phase

File: get_files.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import requests
+import just
+import lxml.html
+import re
+
+# file with one history per line
+with open('history_files.txt', 'w') as g:
+    with open('repos.txt', 'r') as f:
+        for num, link in enumerate(f.readlines()):
```
Comment (vmarkovtsev) [inline get_files.py]: for num, link in enumerate(f):

File: get_files.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            link = link.rstrip()
+            html = requests.get(link).text
```
Comment (vmarkovtsev) [inline get_files.py]: Please use [scrapy](https://scrapy.org/)

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Shell neural completion - Work in Progress
+
+## Phase 1 : Build data pipeline
+
+Gather bash and zsh history dataset
+Clean and Fetch files into one : 'history_files.txt' one history per line
+
+### Initialization
```
Comment (vmarkovtsev) [inline README.md]: Installation

File: library.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+## Useful Papers and Repositories
+
+* https://github.com/rupa/z
+
```
Comment (vmarkovtsev) [inline library.md:4]: Why newline here? If you want to mark this, use smth after * but before the link

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+from setuptools import setup, find_packages
+
+setup(name='shcomplete',
+      version='1.0.0',
+      license="Apache 2.0",
+      author="source{d}",
+      )
```
Comment (vmarkovtsev) [inline setup.py]: Please remove this and add in the next PR. Packaging will be in the future.

File: requirements.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+PyGithub
+requests
+just
+scrapy
```
Comment (vmarkovtsev) [inline requirements.txt]: requirements always have fixed or almost fixed versions. I recommend using fixed versions unless you really know the packages well.

File: shcomplete/__main__.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    subparsers = parser.add_subparsers(help="Commands", dest="command")
+    repos_parser = subparsers.add_parser("repos",
+        help="Return the list of repositories where to find history files")
+    repos_parser.set_defaults(handler=fetch_repos)
+    repos_parser.add_argument(
+        "-i", "--token", required=True,
+         help="Github API token")
+    repos_parser.add_argument(
+        "-o", "--output", required=True,
+        help="path to the output file")
```
Comment (vmarkovtsev) [inline shcomplete/__main__.py]: Please always start help with a capital letter and end with a dot

File: shcomplete/corpus.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import os
+
+class HistoryFile(object):
+
+    def write1file(self, path_to_file, output):
```
Comment (vmarkovtsev) [inline shcomplete/corpus.py]: Better `write_file`

File: shcomplete/corpus.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        """
+        Write path_to_file on 1 line of output file
+        """
+        with open(output, "a") as fout:
+            with open(path_to_file, "r") as f:
+                lines = f.readlines()
```
Comment (vmarkovtsev) [inline shcomplete/corpus.py]:
```
for line in f:
    ...
```

File: shcomplete/corpus.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+                for line in lines:
+                    line = line.rstrip() + " "
+                    fout.write(line)
+                fout.write("\n")
+            f.close()
```
Comment (vmarkovtsev) [inline shcomplete/corpus.py]: Never `close()` if you use `with`. `close()` is used in the following case:

```
f = open(...)
try:
    # work with f
except:
    f.close()
```

`with` is a nice wrapper for this try-except. Further read: https://stackoverflow.com/questions/1984325/explaining-pythons-enter-and-exit

File: shcomplete/corpus.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+def write_corpus(args):
+    """
+    Write all the history files into an output txt file. One file per line.
+    """
+    Nb_file=0
+    for root, d, files in os.walk("shcomplete/data"):
+        for f in files:
+            Nb_file += 1
+            path_to_file = os.path.join(root, f)
+            print("--------------------------")
```
Comment (vmarkovtsev) [inline shcomplete/corpus.py]: logging please

File: shcomplete/corpus.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+                fout.write("\n")
+            f.close()
+        fout.close()
+
+
+def write_corpus(args):
+    """
+    Write all the history files into an output txt file. One file per line.
+    """
+    Nb_file=0
```
Comment (vmarkovtsev) [inline shcomplete/corpus.py]: Style: never start a variable or a method from a capital letter

File: shcomplete/corpus.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    for root, d, files in os.walk("shcomplete/data"):
+        for f in files:
+            Nb_file += 1
+            path_to_file = os.path.join(root, f)
+            print("--------------------------")
+            print("file nº", Nb_file)
+            print(path_to_file)
+            print("--------------------------")
+            write_f = HistoryFile()
+            write_f.write1file(path_to_file, args.output)
```
Comment (vmarkovtsev) [inline shcomplete/corpus.py]: OK, now I see. You don't need a class for this single op, do you?

File: shcomplete/filtering.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
+import re
+import os
+
+class Zsh(object):
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: You seem to have several classes with the same interface. Please extract the interface.

```
class Shell:
    def existing_timestamps(seld, file_name):
        raise NotImplementedError
    ...

class Zsh(Shell):
    ...
```

P.S. Do not inherit from `object`. This is needed to be compatible with python 2.6 and later - we fix to >=3.4

File: shcomplete/filtering.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+    def existing_timestamps(seld, file_name):
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: Always name your methods as verbs (or looking as verbs). `detect_timestamps` or `timestamps_exist` better

File: shcomplete/filtering.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+        """
+        Return 1 if there are timestamps to remove in zsh file_name
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: We are coding in Python, not C/C++ - return a `bool`

File: shcomplete/filtering.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+        """
+        with open(file_name, "r") as f:
+            lines = f.readlines()
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: del

File: shcomplete/filtering.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+            timestamp = r": \d{10}:\d{1,};"
+            test = re.match(timestamp, lines[0])
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: return re.match(timestamp, f.readline())

Use `re.compile` and cache the regexp as a static variable.

```
class Zsh:
    TIMESTAMP_RE = re.compile(<your string here>)

    def ...:
        return self.TIMESTAMP_RE.match(f.readline())
```

File: shcomplete/filtering.py [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+            return test
+
+    def remove_timestamps(self, file_name):
+        with open(file_name, "r+") as f:
+            lines = f.readlines()
+            f.seek(0)
+            f.truncate()
+            timestamp = r": \d{10}:\d{1,};"
+            for line in lines:
+                command = re.split(timestamp, line)
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: same here - reuse the precompiled regexp

File: shcomplete/filtering.py [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+class Fish(object):
+
+    def existing_timestamps(seld, file_name):
+        """
+        Return 1 if there are timestamps to remove in fish file_name
+        """
+        with open(file_name, "r") as f:
+            lines = f.readlines()
+            pattern = r"- cmd: "
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: set this pattern as a static variable

File: shcomplete/filtering.py [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+                    f.write(command[1])
+
+class Fish(object):
+
+    def existing_timestamps(seld, file_name):
+        """
+        Return 1 if there are timestamps to remove in fish file_name
+        """
+        with open(file_name, "r") as f:
+            lines = f.readlines()
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: del

File: shcomplete/filtering.py [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+            pattern = r"- cmd: "
+            test = pattern in lines[0]
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: `readline()`

File: shcomplete/filtering.py [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+            return test
+
+    def remove_timestamps(self, file_name):
+        with open(file_name, "r+") as f:
+            lines = f.readlines()
+            f.seek(0)
+            f.truncate()
+            pattern = r"- cmd: "
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: reuse the static var here

File: shcomplete/filtering.py [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+
+def filter_timestamps(args):
+    """
+    Remove timestamps in all history files that need it 
+    """
+    for root, d, files in os.walk("shcomplete/data"):
+        for f in files:
+            path_to_f = os.path.join(root, f)
+
+            if f.endswith("zsh_history"):
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: Use static vars from the classes.

```
__shells__ = []

def register_shell(cls):
    __shells__.append(cls)
    return cls

class Shell:
    @classmethod
    def matches(cls, file_name):
        return file_name.endswith(cls.HISTORY_FILE)

@register_shell
class Zsh(Shell):
    HISTORY_FILE = "zsh_history"
    ...

...

for f in files:
    for cls in __shells__:
        if cls.matches(f):
            # use the public interface
```

File: shcomplete/filtering.py [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+                zsh_filtering = Zsh()
+                if zsh_filtering.existing_timestamps(path_to_f):
+                    zsh_filtering.remove_timestamps(path_to_f)
+
+            if f.endswith("fish_history"):
+                fish_filtering = Fish()
+                if fish_filtering.existing_timestamps(path_to_f):
+                    fish_filtering.remove_timestamps(path_to_f)
```
Comment (vmarkovtsev) [inline shcomplete/filtering.py]: Where is bash_history BTW?

File: shcomplete/repos.py [hunk 27]
```diff
@@ -0,0 +1,78 @@
+import time
+import os
+import argparse
+import sys
+
+from github import Github as GitHub, RateLimitExceededException
+
+class GithubShell(object):
```
Comment (vmarkovtsev) [inline shcomplete/repos.py]: GitHubShellDiscoverer

File: shcomplete/repos.py [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+
+    def get_repos_list(self):
+        """
+        Get the list of repositories
+        """
+
+        repos_fish = []
+        for num, repo in enumerate(self._api.search_code("filename:fish_history size:>0")):
+            repos_fish.append(repo.html_url)
+            print(num + 1, repos_fish[-1])
```
Comment (vmarkovtsev) [inline shcomplete/repos.py]: logging

File: shcomplete/repos.py [hunk 29]
```diff
@@ -0,0 +1,78 @@
...
+            time.sleep(0.5)
+        repos_fish = {x for x in repos_fish if x.endswith("/fish_history")}
```
Comment (vmarkovtsev) [inline shcomplete/repos.py]: You should not hardcode shells here but rather use HISTORY_FILE from your classes

Comment (vmarkovtsev) [reply shcomplete/repos.py]: And do not duplicate the code three times. Use `__shells__`

File: shcomplete/repos.py [hunk 30]
```diff
@@ -0,0 +1,78 @@
...
+
+        repos_bash = []
+        for num, repo in enumerate(self._api.search_code("filename:bash_history size:>=950 -user:Dahs81")):
+            repos_bash.append(repo.html_url)
+            print(num + 1, repos_bash[-1])
+            time.sleep(0.7)
```
Comment (vmarkovtsev) [inline shcomplete/repos.py]: Hehe, wow! Why do you need this? At least put this into an instance variable (init in `__init__`), at max remove it.

File: shcomplete/repos.py [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+
+        return repos
+
+    def to_raw_urls(self, list_urls):
+        """
+        Transform the links returned by GitHub API, github.com into raw.githubusercontent.com
+        """
+
+        list_raw_urls = []
+        base = "https://raw.githubusercontent.com"
```
Comment (vmarkovtsev) [inline shcomplete/repos.py]: Make it a constant, move to a static var

File: shcomplete/repos.py [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    Fetch the repositories and write the raw links into an output txt file
+    """
+
+    histories = GithubShell(args.token)
+    list_repos = histories.get_repos_list()
+    list_repos = histories.to_raw_urls(list_repos)
+    with open(args.output, 'w') as f:
+        ens_repos = set(list_repos)
+        for rep in list(ens_repos):
```
Comment (vmarkovtsev) [inline shcomplete/repos.py]: You do not need to convert it to list if you just want to iterate. If you want the same order each time, use `sorted`

File: shcomplete/repospider.py [hunk 33]
```diff
@@ -0,0 +1,15 @@
+from scrapy.spiders import Spider
+from scrap.items import ScrapItem
+from scrapy.http import Request
+import just
+
+class MySpider(Spider):
+    name = "github_spider"
+    allowed_domains = ["github.com"]
+    f = open("repos.txt")
```
Comment (vmarkovtsev) [inline shcomplete/repospider.py]: `with open() as f`
~close~

File: shcomplete/scrap/settings.py [hunk 34]
```diff
@@ -0,0 +1,90 @@
...
+# Scrapy settings for scrap project
+#
+# For simplicity, this file contains only settings considered important or
+# commonly used. You can find more settings consulting the documentation:
+#
+#     http://doc.scrapy.org/en/latest/topics/settings.html
+#     http://scrapy.readthedocs.org/en/latest/topics/downloader-middleware.html
+#     http://scrapy.readthedocs.org/en/latest/topics/spider-middleware.html
+
+BOT_NAME = 'scrap'
```
Comment (vmarkovtsev) [inline shcomplete/scrap/settings.py]: Give it a name

File: shcomplete/scrapy.cfg [hunk 35]
```diff
@@ -0,0 +1,11 @@
...
+#
+# For more information about the [deploy] section see:
+# https://scrapyd.readthedocs.org/en/latest/deploy.html
+
+[settings]
+default = scrap.settings
+
+[deploy]
+#url = http://localhost:6800/
+project = scrap
```
Comment (vmarkovtsev) [inline shcomplete/scrapy.cfg]: Give it a name; remove the comments in `*.py` files except `settings.py`

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: shcomplete/filtering.py (hunk 26 above)
Comment (warenlg) [reply shcomplete/filtering.py]: no timestamps, seem to be all already clean

Comment (vmarkovtsev) [reply shcomplete/filtering.py]: https://askubuntu.com/questions/391082/how-to-see-time-stamps-in-bash-history

**Pull request closed without merging**